src/components/Grid.re

```diff
@@ -1,32 +1,83 @@
-  width(`calc((`add, pct(100.0), px(spacing)))),
-  margin(px(- spacing / 2)),
+type flexBasisLength = [
+  length
+  | `auto
+  | `fill
+  | `content
+  | `maxContent
+  | `minContent
+  | `fitContent
 ];
 
-let gridItem = (spacing: int) => [padding(px(spacing / 2))];
+type gridItemWidth =
+  | Columns(int)
+  | Grow
+  | Auto
+  | Flex(float, int, flexBasisLength);
 
-let component = ReasonReact.statelessComponent("Grid");
+type containerWrap = [ | `nowrap | `wrap | `wrapReverse];
 
 type gridType =
-  | Container
-  | Item;
+  | Container(containerWrap)
+  | Item(gridItemWidth);
 
-[@react.component]
-let make = (~type_: gridType, ~children) => {
-  let getGridStyles = (): list(rule) => {
-    switch (type_) {
-    | Container =>
-      [display(`flex)]
-      @ gridContainer(Theme.spacingPx.small)
-      @ [media(Breakpoints.up(Md), gridContainer(Theme.spacingPx.medium))]
-
-    | Item =>
-      [flexGrow(1.0), flexShrink(1), flexBasis(auto)]
-      @ gridItem(Theme.spacingPx.small)
-      @ [media(Breakpoints.up(Md), gridItem(Theme.spacingPx.medium))]
+module Styles = {
+  let containerSpacing = (spacing: int) => [
+    width(`calc((`add, pct(100.0), px(spacing)))),
+    margin(px(- spacing / 2)),
+  ];
+  let itemSpacing = (spacing: int) => [padding(px(spacing / 2))];
+
+  let calculateColumnsWidth = (columns: int) =>
+    float_of_int(columns) /. 12.0 *. 100.0;
+
+  let getFlexStyles = (grow, shrink, basis) => [
+    flexGrow(grow),
+    flexShrink(shrink),
+    flexBasis(basis),
+  ];
+
+  let gridItemWidthStyle = (width: gridItemWidth) =>
+    switch (width) {
+    | Columns(columns) =>
+      let widthPct = calculateColumnsWidth(columns);
+
+      getFlexStyles(0.0, 1, pct(widthPct)) @ [maxWidth(pct(widthPct))];
+    | Flex(grow, shrink, basis) => getFlexStyles(grow, shrink, basis)
+    | Grow => getFlexStyles(1.0, 1, auto)
+    | Auto => []
     };
+
+  let gridContainer = (wrapValue: containerWrap) =>
+    style(
+      [display(`flex), flexWrap(wrapValue), boxSizing(borderBox)]
+      @ containerSpacing(Theme.spacingPx.small)
+      @ [
+        media(Breakpoints.up(Md), containerSpacing(Theme.spacingPx.medium)),
+      ],
+    );
+
+  let gridItem = (itemWidth: gridItemWidth) =>
+    style(
+      [boxSizing(borderBox)]
+      @ gridItemWidthStyle(itemWidth)
+      @ itemSpacing(Theme.spacingPx.small)
+      @ [media(Breakpoints.up(Md), itemSpacing(Theme.spacingPx.medium))],
+    );
+};
+
+let getGridStyles = (type_: gridType) => {
+  switch (type_) {
+  | Container(wrapValue) => Styles.gridContainer(wrapValue)
+  | Item(width) => Styles.gridItem(width)
   };
-  <div className={Css.style(getGridStyles())}> children </div>;
+};
+
+[@react.component]
+let make = (~type_: gridType, ~children, ~className=?) => {
+  let gridClassName = getGridStyles(type_);
+  let customClassName = Belt.Option.getWithDefault(className, "");
+
+  <div className={j|$gridClassName $customClassName|j}> children </div>;
```

Comment:
What's happening here? Is it JS interop? Why is it necessary?

Reply:
It is a string interpolation.